### PR TITLE
bpo-17611: Move unwinding of stack from interpreter to compiler

### PR DESCRIPTION
**Pull request by nascheme**
This is a re-base and slightly more polished version of PR #2827.  I think this change is a good idea as it reduces the size of the stacked used by Python frames.  

Also, it makes the bytecode have an "ahead of time" computable effect on the stack.  See the changes in PyCompile_OpcodeStackEffect().  E.g. before WITH_CLEANUP_FINISH had a "XXX" comment saying it dropped at least one stack item but at runtime maybe more.  After this change, WITH_CLEANUP_FINISH always drops 6 items. 

I understand that is important for tools that try to JIT compile Python bytecode.  Having an ahead of time computable stack effect makes the JIT a lot simpler.

https://bugs.python.org/issue17611

### CODE REVIEW

**Comment by rhettinger** [comment]
+1 Overall, this looks like a net win.

File: Lib/test/test_peepholer.py
Comment (ncoghlan) [inline Lib/test/test_peepholer.py]: Is this truly non-deterministic now? Or should the expected number of return opcodes be modified instead of making the check more permissive?

File: Lib/test/test_peepholer.py
Comment (ncoghlan) [inline Lib/test/test_peepholer.py]: As for previous comment - can we adjust the number of expected returns here instead of relaxing the check?

File: Python/ceval.c
Comment (ncoghlan) [inline Python/ceval.c]: These should be regular validity checks, rather than assertions - bytecode editing may inject a `RERAISE` at a point where there isn't a valid exception state on top of the stack, and that should give a Python exception rather than a C level assertion failure.

File: Python/ceval.c
Comment (ncoghlan) [inline Python/ceval.c]: As above - this should be a check that raises `PyExc_SystemError`, not a C level assertion.

**Review by ncoghlan — DISMISSED** [review]
Reviewing the eval loop changes, my main comment is that we can't safely make assertions about the expected eval stack state when a particular bytecode is executed, since synthetically constructed bytecode may fail to meet those expectations in weird and exciting ways.

Instead, we should use explicit expectation checks, and raise `SystemError` when they're not met.

(Note: I haven't reviewed the compiler changes yet)

File: Lib/test/test_peepholer.py
Comment (nascheme) [reply Lib/test/test_peepholer.py]: I looks to me like that check got changed in an old version of the patch and maybe isn't needed.  I'm going to review all the unit test changes and see if they are still necessary with the final version of the patch.  Antoine went through quite a few revisions.

Comment (nascheme) [reply Lib/test/test_peepholer.py]: I believe this part of the patch can be reverted.  self.assertEqual(len(returns), 6) works and I see no reason why the number of return statements should be indeterminate.

Comment (nascheme) [reply Lib/test/test_peepholer.py]: Sorry, the change below, "len(returns), 2", is the one that can be reverted. This one now returns 5.  I believe the reason is that there is no longer a SETUP_LOOP opcode and the peephole op is now eliminating (correctly, as far as I see), the "return 6" at the end of the function body.  So, should always be 5 return statements in bytecode now.

File: Python/ceval.c
Comment (nascheme) [reply Python/ceval.c]: I'm surprised to hear that.  Do we really expect ceval to be resistant to crashing if fed arbitrary bytecode?  That seems a tough standard to meet and would seem to imply a lot of overhead in checking that the bytecode doesn't do anything unsafe (e.g. re whats on the stack). If that is the case, I expect this patch needs a lot of fixing yet.

Comment (ncoghlan) [reply Python/ceval.c]: We don't always check directly in ceval - we'll often rely on the fact that ceval calls public CPython C APIs, and those already have their own error checking.

But we do have a lot of checks for "impossible" situations (search for SystemError in ceval.c), and the general principle of "bytecode hacking may cause an exception, but it won't crash the process" applies. (That said, there are some existing `assert` calls related to the stack state for function calls, so this could be worth raising on python-dev)

**Comment by nascheme** [comment]
I did a fast benchmark using pyperformance.  Results are [here](https://gist.github.com/nascheme/06d39ecebe7f7d39941161afe9cd764a).  Roughly 2% faster so I would conclude that performance did not get worse.

File: Python/ceval.c
Comment (nascheme) [reply Python/ceval.c]: What kinds of bytecode changes are allowed?  I wrote a very simple minded fuzzer for ceval, giving it randomly generated code.  It crashes Python 3.6 immediately.  I don't disagree with better checking in principle but it seems a very high standard to meet without having some significant performance affects.

Comment (ncoghlan) [reply Python/ceval.c]: Yeah, given the existing assert() checks for function calls, I'll withdraw this comment for now.

**Comment by pitrou** [comment]
> Having an ahead of time computable stack effect makes the JIT a lot simpler.

I wouldn't say "a lot", since bytecode parsing isn't the most complex part of a Python JIT, but I definitely think this is a welcome change (fixing Numba's handling of bytecode stack effect took a bit too much time for my taste :-)).

**Comment by serhiy-storchaka** [comment]
See also other continuation of Antoine's PR: https://github.com/serhiy-storchaka/cpython/commits/unwind_stack .

I'm working on slightly different approach to this issue (not published yet).

**Comment by nascheme** [comment]
This following function (simplified from gevent threadpool.py) causes the compiler to crash::

	def func():
	    try:
		try:
		    func()
		except:
		    return
		finally:
		    pass
	    finally:
		return

I'm investigating.  Seems to infinite recursion in compiler_unwind_fblock().

**Comment by nascheme** [comment]
I have an even simpler example::

	def func():
	    try:
		try:
		    True
		except:
		    return
	    finally:
		return

The fb_unwind function for the HANDLER_CLEANUP fblocktype seems to be buggy.
Still trying to figure out how this all works.

**Comment by nascheme** [comment]
The fblock unwind logic in finally bodies is still buggy.  It unwinds all blocks, not just the blocks enclosing the finally body.  I think the fix is not too difficult: allocate fblockinfo on the C stack, have an enclosing block pointer in the fblockinfo struct.  When unwind happens, walk the enclosing block pointers.  fblock_unwind_finally_try() will have to set the current fblockinfo in the 'c' struct so that unwinds coming from VISIT_SEQ()  will unwind from the correct place.

As I mended in the issue tracker, I think the finally body code duplication is the correct approach.  I spent a good part of Sunday studying the patch.  The ceval/bytecode changes look pretty solid.  The compiler needs some fixing yet.

**Comment by nascheme** [comment]
All my tests are passing now.  Still could use some polish before merging.

File: Lib/test/test_compile.py
Comment (pitrou) [inline Lib/test/test_compile.py:761]: Do you have any idea how to solve this one?

**Comment by nascheme** [comment]
This PR still has issues with frame.set_lineno() and they don't appear to be trivial to fix.  I'm closing the PR.  Mark Shannon, who was the original author of this patch, has expressed interested in trying to resolve the issues.  So, I think it best to leave it to him.

**Pull request closed without merging**
